drv/i2c-devices/src/emc2305.rs

```diff
@@ -0,0 +1,334 @@
+    Fan5MinDrive = 0x78,
+    Fan5ValidTach = 0x79,
+    Fan5DriveFailBandLo = 0x7a,
+    Fan5DriveFailBandHi = 0x7b,
+    Tach5TargetLo = 0x7c,
+    Tach5TargetHi = 0x7d,
+    Tach5ReadingHi = 0x7e,
+    Tach5ReadingLo = 0x7f,
+
+    SoftwareLock = 0xef,
+    ProductFeatures = 0xfc,
+    ProductId = 0xfd,
+    MfgId = 0xfe,
+    Revision = 0xff,
+}
+
+pub struct Emc2305 {
+    pub device: I2cDevice,
+}
+
+pub const MAX_FANS: u8 = 5;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct Fan(u8);
+
+impl TryFrom<u8> for Fan {
+    type Error = ();
+    /// Fans are based on a 0-based index. This should *not* be the number
+    /// of the fan (the fan numbers have a 1-based index)
+    fn try_from(index: u8) -> Result<Self, Self::Error> {
+        if index >= MAX_FANS {
+            Err(())
+        } else {
+            Ok(Self(index))
+        }
+    }
+}
+
+impl Fan {
+    fn register(&self, base: Register) -> Register {
+        Register::from_u8((base as u8) + self.0 * 0x10).unwrap()
+    }
+
+    fn configuration1(&self) -> Register {
+        self.register(Register::Fan1Cfg1)
+    }
+
+    fn configuration2(&self) -> Register {
+        self.register(Register::Fan1Cfg2)
+    }
+
+    fn tach_count(&self) -> Register {
+        self.register(Register::Tach1ReadingHi)
+    }
+
+    fn pwm_target(&self) -> Register {
+        self.register(Register::Fan1Setting)
+    }
+}
+
+fn read_reg8(
+    device: &I2cDevice,
+    register: Register,
+) -> Result<u8, ResponseCode> {
+    device.read_reg::<u8, u8>(register as u8)
+}
+
+fn read_reg16(
+    device: &I2cDevice,
+    register: Register,
+) -> Result<[u8; 2], ResponseCode> {
+    device.read_reg::<u8, [u8; 2]>(register as u8)
+}
+
+fn write_reg8(
+    device: &I2cDevice,
+    register: Register,
+    val: u8,
+) -> Result<(), ResponseCode> {
+    device.write(&[register as u8, val])
+}
+
+fn write_reg16(
+    device: &I2cDevice,
+    register: Register,
+    val: u16,
+) -> Result<(), ResponseCode> {
+    device.write(&[register as u8, (val >> 8) as u8, (val & 0xff) as u8])
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    ZeroTach(Fan),
+    TachOverflow(u32),
+    BadFanCount(u8),
+    None,
+}
+
+ringbuf!(Trace, 6, Trace::None);
+
+impl Emc2305 {
+    pub fn new(device: &I2cDevice) -> Self {
+        Self { device: *device }
+    }
+
+    pub fn initialize(&self, fan_count: u8) -> Result<(), ResponseCode> {
+        if fan_count > MAX_FANS {
+            ringbuf_entry!(Trace::BadFanCount(fan_count));
+            return Err(ResponseCode::BadArg);
+        }
+
+        assert!(fan_count <= MAX_FANS);
+        let device = &self.device;
+
+        // Enable the watchdog at all times
+        let mut config =
+            Configuration(read_reg8(device, Register::Configuration)?);
+        config.set_watchdog_enable(true);
+        write_reg8(device, Register::Configuration, config.0)?;
+
+        for fan in 0..fan_count {
+            let fan = Fan::try_from(fan).unwrap();
```

Comment:
nit: should this be `unwrap_lite()`?

Reply:
Sure (and fixed the other `unwrap` in here as well)